### PR TITLE
1 second delay during server creation

### PR DESCRIPTION
**Pull request by manishtomar**
This is done by limiting server creation to 1 server at a time and injecting 1 second delay from create server response.

### CODE REVIEW

**Comment by lvh** [comment]
This closes #1221, right?

File: otter/worker/launch_server_v1.py [hunk 1]
```diff
@@ -308,6 +308,13 @@ def _check_server_created(f):
         d.addBoth(_check_results, f)
         return d
 
+    def _create_with_delay():
+        d = _treq.post(path, headers=headers(auth_token),
+                       data=json.dumps({'server': server_config}), log=log)
+        # Add 1 second delay to space 1 second between server creations
+        d.addCallback(delay, clock, 1)
```
Comment (lvh) [inline otter/worker/launch_server_v1.py]: It is a bit unfortunate that this gets the result, and then sits on it for a second. We need to space outgoing requests, but you can still asynchronously space results without having to wait on them here.

File: otter/test/worker/test_launch_server_v1.py [hunk 2]
```diff
@@ -1156,24 +1163,37 @@ def test_create_server_limits(self):
...
         }
 
-        ret_ds = [create_server('http://url/', 'my-auth-token', server_config)
+        ret_ds = [create_server('http://url/', 'my-auth-token',
+                                server_config, clock=self.clock)
                   for i in range(3)]
 
-        # no result in any of them and only first 2 treq.post is called
+        # no result in any of them and only 1 treq.post is called
         [self.assertNoResult(d) for d in ret_ds]
```
Comment (lvh) [inline otter/test/worker/test_launch_server_v1.py]: I realize you didn't write this in this PR, but side-effectful LCs are super gross and we should probably turn that into a for loop :)

File: otter/worker/launch_server_v1.py (hunk 1 above)
Comment (manishtomar) [inline otter/worker/launch_server_v1.py]: This is the easiest way I thought of throttling requests on a single node. Another way is to space out job creation in [supervisor](https://github.com/rackerlabs/otter/blob/master/otter/supervisor.py#L559). Is this what you were suggesting? This is probably simpler but doesn't ensure spacing when multiple policies (of possibly different groups) are getting executed at same time. I can understand if that is considered an over-optimization at this point.

Comment (lvh) [inline otter/worker/launch_server_v1.py]: I don't think that this should block this PR right now, FWIW :-)

There are two ways to do this:

- Supervisor, as you were suggesting. That would space out *jobs*, not server creations; not quite what Compute is asking for, and, in the problematic cases, not something that actually performs their experiment.
- Separate the concern of "one at a time" from "at most one per second". I.e., don't change the thing you're changing here at all; that fn creates a server, and it still creates a server (no added delay). Instead, Make the delay be about adding things to the `DeferredSemaphore` in the first place, by e.g. putting create server requests in a queue, and consuming them every *timeout* seconds.

Comment (manishtomar) [inline otter/worker/launch_server_v1.py]: That does sound much cleaner way to do it. It makes its intent very clear. I got stuck in current code base and thought from that angle :). As you suggested, this doesn't need to block this PR now. Instead the idea can be utilized in converger.

**Comment by lvh** [comment]
LGTM in the sense that it does address #1221. I am concerned that with the number of servers we create (and considering the Hurst exponent in that process), the one second delay is going to cause some pretty severe degradation; but that's part of the exercise, I guess.

**Comment by manishtomar** [comment]
This addresses #1221 in worker only. That ticket is for converger. I think we should keep that open until it is implemented in converger.

**Comment by manishtomar** [comment]
jenkins please run integration tests.

**Comment by lvh** [comment]
@manishtomar Argh; did that change? We definitely gave out that ticket to a bunch of other people who are waiting for it to happen in worker right now, so I'll make a separate ticket for doing it in convergence.

**Comment by lvh** [comment]
Assuming the tests pass, this LGTM. I would like to see if we can get the "what does #1221 even mean" issue resolved before merging it, but that's not a blocker.

**Comment by lvh** [comment]
Huh; where did the build info go?

**Comment by lvh** [comment]
Never mind, Github's auto-update went a little wonky :)

**Pull request merged**